runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -5429,3 +5620,78 @@ MM_CopyForwardScheme::randomDecideForceNonEvacuatedRegion(UDATA ratio) {
 	return ret;
 }
 
+MMINLINE bool
+MM_CopyForwardScheme::isCompressedSurvivor(void *heapAddr)
+{
+	UDATA compressedCardOffset = ((UDATA)heapAddr - (UDATA)_heapBase) / CARD_SIZE;
+	UDATA compressedCardIndex = compressedCardOffset / COMPRESSED_CARDS_PER_WORD;
+	UDATA compressedSurvivorWord = _compressedSurvivorTable[compressedCardIndex];
+	bool isSurvivor = false;
+
+	if (AllCompressedCardsInWordClean != compressedSurvivorWord) {
+		UDATA bit = compressedCardOffset % COMPRESSED_CARDS_PER_WORD;
+		isSurvivor = (CompressedCardSurvivor == ((compressedSurvivorWord >> bit) & 1));
+	}
+	return isSurvivor;
+}
+
+MMINLINE void
+MM_CopyForwardScheme::setCompressedSurvivorCards(MM_EnvironmentVLHGC *env, void *startHeapAddress, void *endHeapAddress)
+{
+	UDATA compressedCardStartOffset = ((UDATA)startHeapAddress - (UDATA)_heapBase) / CARD_SIZE;
+	UDATA compressedCardStartIndex = compressedCardStartOffset / COMPRESSED_CARDS_PER_WORD;
+	UDATA compressedCardEndOffset = (((UDATA)endHeapAddress - (UDATA)_heapBase) + (CARD_SIZE -1))/ CARD_SIZE;
+	UDATA compressedCardEndIndex = compressedCardEndOffset / COMPRESSED_CARDS_PER_WORD;
+	UDATA mask = 1;
+	UDATA endOfWord = ((UDATA)1) << (COMPRESSED_CARDS_PER_WORD - 1);
+	UDATA compressedSurvivorWord = AllCompressedCardsInWordClean;
+
+	UDATA *compressedSurvivor = &_compressedSurvivorTable[compressedCardStartIndex];
+
+	UDATA shiftStart = compressedCardStartOffset % COMPRESSED_CARDS_PER_WORD;
+	mask = mask << shiftStart;
+	UDATA offset = compressedCardStartOffset;
+	UDATA idx = compressedCardStartIndex;
+	if (idx == compressedCardEndIndex) {
+		endOfWord = ((UDATA)1) << ((compressedCardEndOffset - 1) % COMPRESSED_CARDS_PER_WORD);
+	}
+//	PORT_ACCESS_FROM_ENVIRONMENT(env);
+//	j9tty_printf(PORTLIB, "setCompressedSurvivorCards startHeapAddress=%p, endHeapAddress=%p, compressedCardStartOffset=%zu, compressedCardEndOffset=%zu, compressedCardStartIndex=%zu, compressedCardEndIndex=%zu, endOfWord=%p, mask=%p\n",
+//			startHeapAddress, endHeapAddress, compressedCardStartOffset, compressedCardEndOffset, compressedCardStartIndex, compressedCardEndIndex, endOfWord, mask);
+	while (offset < compressedCardEndOffset) {
```

Comment:
If I calculated this correctly, TLH of 128KB would result in 4-5 atomic updates. It's not bad, while theoretically it should be only 2 atomics at far ends, and non-atomic for intermediate ones). I'm ok to leave the code simpler (if it's not very simple to recognize the first and last word), but at least please propagate this observation down to code (as a comment).